mvn clean quarkus:dev not working in mixed Kotlin project

**Describe the bug**
Project with mixed Java/Kotlin doesn't compile after moving to `quarkus-maven-plugin:1.2.0.Final`.  Either the Kotlin compiler isn't running at all or it it is not running before Java (as required).

Switching the plugin version back to `1.1.1.Final` the project builds/runs correctly again.

**Expected behavior**
Using mixed Java/Kotlin with Quarkus builds correctly.

**Actual behavior**
Java source files cannot find symbols defined in Kotlin source files and produces compilation errors.

**To Reproduce**
1.  Create a mixed Java/Kotlin project built with Maven.
2. Reference a symbol defined in a Kotlin file from a Java file.
3. Build (e.g. using `package`) or run (e.g. using `quarks:dev`).

**Configuration**
N/A

**Screenshots**
N/A

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
   Darwin ... 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
 - Output of `java -version`:
   java version "11.0.6" 2020-01-14 LTS
   Java(TM) SE Runtime Environment 18.9 (build 11.0.6+8-LTS)
   Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.6+8-LTS, mixed mode)
 - Output of `java -version`: 
N/A
 - GraalVM version (if different from Java): 
N/A
 - Quarkus version or git rev: 
`1.2.0.Final`

**Additional context**
N/A

To fix this we would need to also take the kotlin-maven-plugin into account in `DevMojo` when a compilation is detected.
However it's not straightforward since it seems some setup is needed otherwise we get:

```
Caused by: org.codehaus.plexus.component.repository.exception.ComponentLookupException: java.util.NoSuchElementException
      role: org.jetbrains.kotlin.maven.KotlinMavenPluginExtension
  roleHint: all-open
    at org.codehaus.plexus.DefaultPlexusContainer.lookup (DefaultPlexusContainer.java:267)
    at org.codehaus.plexus.DefaultPlexusContainer.lookup (DefaultPlexusContainer.java:255)
    at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.loadCompilerPlugins (KotlinCompileMojoBase.java:321)
    at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.getCompilerPluginOptions (KotlinCompileMojoBase.java:349)
    at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.configureCompilerArguments (KotlinCompileMojoBase.java:489)
    at org.jetbrains.kotlin.maven.KotlinCompileMojoBase.execute (KotlinCompileMojoBase.java:215)
    at org.jetbrains.kotlin.maven.K2JVMCompileMojo.execute (K2JVMCompileMojo.java:222)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:119)
    at io.quarkus.maven.DevMojo.executeCompileGoal (DevMojo.java:369)
```

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -237,6 +237,10 @@ public class DevMojo extends AbstractMojo {
     public void execute() throws MojoFailureException, MojoExecutionException {
 
         mavenVersionEnforcer.ensureMavenVersion(getLog(), session);
+
+        //we always want to compile if needed, so if it is run from the parent it will compile dependent projects
+        handleAutoCompile();
+
         Plugin pluginDef = MojoUtils.checkProjectForMavenBuildPlugin(project);
 
         if (pluginDef == null) {
@@ -248,34 +252,6 @@ public class DevMojo extends AbstractMojo {
 
         if (!sourceDir.isDirectory()) {
             getLog().warn("The project's sources directory does not exist " + sourceDir);
-        }
-        //we check to see if there was a compile (or later) goal before this plugin
-        boolean compileNeeded = true;
-        for (String goal : session.getGoals()) {
-            if (POST_COMPILE_PHASES.contains(goal)) {
-                compileNeeded = false;
-                break;
-            }
-            if (goal.endsWith("quarkus:dev")) {
-                break;
-            }
-        }
-
-        //if the user did not compile we run it for them
-        if (compileNeeded) {
-            // compile the Kotlin sources if needed
-            final String kotlinMavenPluginKey = ORG_JETBRAINS_KOTLIN + ":" + KOTLIN_MAVEN_PLUGIN;
-            final Plugin kotlinMavenPlugin = project.getPlugin(kotlinMavenPluginKey);
-            if (kotlinMavenPlugin != null) {
-                executeCompileGoal(kotlinMavenPlugin, ORG_JETBRAINS_KOTLIN, KOTLIN_MAVEN_PLUGIN);
-            }
-
-            // Compile the Java sources if needed
-            final String compilerPluginKey = ORG_APACHE_MAVEN_PLUGINS + ":" + MAVEN_COMPILER_PLUGIN;
-            final Plugin compilerPlugin = project.getPlugin(compilerPluginKey);
-            if (compilerPlugin != null) {
-                executeCompileGoal(compilerPlugin, ORG_APACHE_MAVEN_PLUGINS, MAVEN_COMPILER_PLUGIN);
-            }
         }
 
         try {
@@ -358,6 +334,37 @@ public class DevMojo extends AbstractMojo {
 
         } catch (Exception e) {
             throw new MojoFailureException("Failed to run", e);
+        }
+    }
+
+    private void handleAutoCompile() throws MojoExecutionException {
+        //we check to see if there was a compile (or later) goal before this plugin
+        boolean compileNeeded = true;
+        for (String goal : session.getGoals()) {
+            if (POST_COMPILE_PHASES.contains(goal)) {
+                compileNeeded = false;
+                break;
+            }
+            if (goal.endsWith("quarkus:dev")) {
+                break;
+            }
+        }
+
+        //if the user did not compile we run it for them
+        if (compileNeeded) {
+            // compile the Kotlin sources if needed
+            final String kotlinMavenPluginKey = ORG_JETBRAINS_KOTLIN + ":" + KOTLIN_MAVEN_PLUGIN;
+            final Plugin kotlinMavenPlugin = project.getPlugin(kotlinMavenPluginKey);
+            if (kotlinMavenPlugin != null) {
+                executeCompileGoal(kotlinMavenPlugin, ORG_JETBRAINS_KOTLIN, KOTLIN_MAVEN_PLUGIN);
+            }
+
+            // Compile the Java sources if needed
+            final String compilerPluginKey = ORG_APACHE_MAVEN_PLUGINS + ":" + MAVEN_COMPILER_PLUGIN;
+            final Plugin compilerPlugin = project.getPlugin(compilerPluginKey);
+            if (compilerPlugin != null) {
+                executeCompileGoal(compilerPlugin, ORG_APACHE_MAVEN_PLUGINS, MAVEN_COMPILER_PLUGIN);
+            }
         }
     }
 

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -100,6 +100,9 @@ public class DevMojo extends AbstractMojo {
     private static final String ORG_APACHE_MAVEN_PLUGINS = "org.apache.maven.plugins";
     private static final String MAVEN_COMPILER_PLUGIN = "maven-compiler-plugin";
 
+    private static final String ORG_JETBRAINS_KOTLIN = "org.jetbrains.kotlin";
+    private static final String KOTLIN_MAVEN_PLUGIN = "kotlin-maven-plugin";
+
     /**
      * The directory for compiled classes.
      */
@@ -260,27 +263,19 @@ public class DevMojo extends AbstractMojo {
 
         //if the user did not compile we run it for them
         if (compileNeeded) {
-            // Compile the project
-            final String key = ORG_APACHE_MAVEN_PLUGINS + ":" + MAVEN_COMPILER_PLUGIN;
-            final Plugin plugin = project.getPlugin(key);
-            if (plugin == null) {
-                throw new MojoExecutionException("Failed to locate " + key + " among the project plugins");
+            // compile the Kotlin sources if needed
+            final String kotlinMavenPluginKey = ORG_JETBRAINS_KOTLIN + ":" + KOTLIN_MAVEN_PLUGIN;
+            final Plugin kotlinMavenPlugin = project.getPlugin(kotlinMavenPluginKey);
+            if (kotlinMavenPlugin != null) {
+                executeCompileGoal(kotlinMavenPlugin, ORG_JETBRAINS_KOTLIN, KOTLIN_MAVEN_PLUGIN);
             }
-            Xpp3Dom configuration = (Xpp3Dom) plugin.getConfiguration();
-            if (configuration == null) {
-                configuration = MojoExecutor.configuration();
+
+            // Compile the Java sources if needed
+            final String compilerPluginKey = ORG_APACHE_MAVEN_PLUGINS + ":" + MAVEN_COMPILER_PLUGIN;
+            final Plugin compilerPlugin = project.getPlugin(compilerPluginKey);
+            if (compilerPlugin != null) {
+                executeCompileGoal(compilerPlugin, ORG_APACHE_MAVEN_PLUGINS, MAVEN_COMPILER_PLUGIN);
             }
-            MojoExecutor.executeMojo(
-                    MojoExecutor.plugin(
-                            MojoExecutor.groupId(ORG_APACHE_MAVEN_PLUGINS),
-                            MojoExecutor.artifactId(MAVEN_COMPILER_PLUGIN),
-                            MojoExecutor.version(plugin.getVersion())),
-                    MojoExecutor.goal("compile"),
-                    configuration,
-                    MojoExecutor.executionEnvironment(
-                            project,
-                            session,
-                            pluginManager));
         }
 
         try {
@@ -364,6 +359,25 @@ public class DevMojo extends AbstractMojo {
         } catch (Exception e) {
             throw new MojoFailureException("Failed to run", e);
         }
+    }
+
+    private void executeCompileGoal(Plugin plugin, String groupId, String artifactId) throws MojoExecutionException {
+        Xpp3Dom configuration = (Xpp3Dom) plugin.getConfiguration();
+        if (configuration == null) {
+            configuration = MojoExecutor.configuration();
+        }
+        MojoExecutor.executeMojo(
+                MojoExecutor.plugin(
+                        MojoExecutor.groupId(groupId),
+                        MojoExecutor.artifactId(artifactId),
+                        MojoExecutor.version(plugin.getVersion()),
+                        plugin.getDependencies()),
+                MojoExecutor.goal("compile"),
+                configuration,
+                MojoExecutor.executionEnvironment(
+                        project,
+                        session,
+                        pluginManager));
     }
 
     private Map<Path, Long> readPomFileTimestamps(DevModeRunner runner) throws IOException {

--- a/integration-tests/kogito-maven/src/test/java/io/quarkus/kogito/maven/it/KogitoDevModeIT.java
+++ b/integration-tests/kogito-maven/src/test/java/io/quarkus/kogito/maven/it/KogitoDevModeIT.java
@@ -15,7 +15,7 @@ public class KogitoDevModeIT extends RunAndCheckMojoTestBase {
     @Test
     public void testThatTheKogitoApplicationRuns() throws MavenInvocationException, IOException {
         testDir = getTargetDir("projects/simple-kogito");
-        run("-e");
+        run(false, "-e");
 
         await()
                 .pollDelay(1, TimeUnit.SECONDS)

--- a/integration-tests/kotlin/src/test/java/io/quarkus/kotlin/maven/it/KotlinDevModeIT.java
+++ b/integration-tests/kotlin/src/test/java/io/quarkus/kotlin/maven/it/KotlinDevModeIT.java
@@ -19,7 +19,7 @@ public class KotlinDevModeIT extends RunAndCheckMojoTestBase {
     @Test
     public void testThatTheApplicationIsReloadedOnKotlinChange() throws MavenInvocationException, IOException {
         testDir = initProject("projects/classic-kotlin", "projects/project-classic-run-kotlin-change");
-        runAndCheck();
+        runAndCheck(false);
 
         // Edit the "Hello" message.
         File jaxRsResource = new File(testDir, "src/main/kotlin/org/acme/HelloResource.kt");

--- a/integration-tests/maven/src/test/java/io/quarkus/maven/it/DevMojoIT.java
+++ b/integration-tests/maven/src/test/java/io/quarkus/maven/it/DevMojoIT.java
@@ -53,7 +53,7 @@ public class DevMojoIT extends RunAndCheckMojoTestBase {
     @Test
     public void testThatResteasyWithoutUndertowCanRun() throws MavenInvocationException, IOException {
         testDir = initProject("projects/classic-no-undertow", "projects/project-classic-no-undertow-run");
-        run();
+        run(false);
 
         //make sure that a simple HTTP GET request always works
         IntStream.range(0, 10).forEach(i -> {
@@ -286,7 +286,7 @@ public class DevMojoIT extends RunAndCheckMojoTestBase {
     @Test
     public void testSourceModificationBeforeFirstCallWorks() throws MavenInvocationException, IOException {
         testDir = initProject("projects/classic", "projects/project-classic-source-modification-before-first-call");
-        run();
+        run(true);
 
         File source = new File(testDir, "src/main/java/org/acme/HelloResource.java");
         // Edit the "Hello" message and provide a random string.
@@ -384,7 +384,7 @@ public class DevMojoIT extends RunAndCheckMojoTestBase {
                 .pollInterval(1, TimeUnit.SECONDS)
                 .until(configurationFile::isFile);
 
-        run();
+        run(true);
 
         // Wait until we get "uuid"
         await()

--- a/test-framework/maven/src/main/java/io/quarkus/maven/it/RunAndCheckMojoTestBase.java
+++ b/test-framework/maven/src/main/java/io/quarkus/maven/it/RunAndCheckMojoTestBase.java
@@ -29,11 +29,13 @@ public class RunAndCheckMojoTestBase extends MojoTestBase {
         awaitUntilServerDown();
     }
 
-    protected void run(String... options) throws FileNotFoundException, MavenInvocationException {
+    protected void run(boolean performCompile, String... options) throws FileNotFoundException, MavenInvocationException {
         assertThat(testDir).isDirectory();
         running = new RunningInvoker(testDir, false);
         final List<String> args = new ArrayList<>(2 + options.length);
-        args.add("compile");
+        if (performCompile) {
+            args.add("compile");
+        }
         args.add("quarkus:dev");
         boolean hasDebugOptions = false;
         for (String option : options) {
@@ -56,7 +58,12 @@ public class RunAndCheckMojoTestBase extends MojoTestBase {
     }
 
     protected void runAndCheck(String... options) throws FileNotFoundException, MavenInvocationException {
-        run(options);
+        runAndCheck(true, options);
+    }
+
+    protected void runAndCheck(boolean performCompile, String... options)
+            throws FileNotFoundException, MavenInvocationException {
+        run(performCompile, options);
 
         String resp = getHttpResponse();
 

